.github/pull_request_template.md

```diff
@@ -0,0 +1,53 @@
+Fixes #XXXX.
+
+(Replace XXXX with the id of the issue fixed in this PR. Remove this line if there is no corresponding
+issue. Don't reference the issue in the title of this pull-request.)
+
+Add tags to your PR if you are a committer (only committers have the right to add tags). Add [Design Review] tag
+if this PR should better be reviewed by at least two people.
+Don't forget to add the following tags (if applicable): [Incompatible], [Release Notes], [Compatibility], [Security],
+[Development Blocker]. Add at least one [Area - ] tag, consider creating a new one if none of the existing [Area - ]
+tags is applicable.
+
+### Description
+
+Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's
+not necessary to repeat the description here, however, you may choose to keep one summary sentence.
+
+Describe your patch: what did you change in code? How did you fix the problem?
+
+If there are several relatively logically separate changes in this PR, list them. For example:
+ - Fixed the bug ...
+ - Renamed the class ...
+ - Added a forbidden-apis entry ...
+
+Some of the aspects mentioned above may be omitted for simple and small PRs.
+
+### Design
+
+Please describe any design decisions made, including:
+ - Choice of algorithms
+ - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such
+   as when insufficient resources are available?
+ - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
+ - Method organization and design (how the logic is split between methods, parameters and return types)
+ - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
+
+In addition, describe _at least one_ alternative design (or mention alternative name) for every design (or naming)
+decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen).
+
+If you already did this in the associated issue (e. g. a "Proposal" issue), leave the following sentence:
+
+Design of this change is discussed [here](<link to Github issue or comment where you discuss the design>).
+
+This section may be omitted for really simple and small patches. However, any patch that adds a new production class
+almost certainly shouldn't omit this section.
+
+<hr>
+
+I've self-reviewed this PR (including using the [concurrency checklist](
```

Comment:
I don't think this is necessary. Providing a list of documents should be enough.

Reply:
This is definitely provocative and moves contributors out of their (non-self-reviewing) comfort zone, but this is the point. If there is just text "You can review your PR using this checklists: ...", I'm afraid, barely anybody will  actually do this.

What do you think about the following wording?
```
### Self-review

(Did you self-review your PR?

Additionally, if your PR has any relation to concurrency (...), please use [this checklist](...).)
```

Reply:
Note that I'm not acting as an exemplar here: in fact, I've only self-reviewed *several* (at best) my PRs into Druid, out of 150+ PRs, despite putting together the checklist. I've only once really carefully reviewed an already existing class with the checklist ([that resulted in finding a bunch of bugs](https://github.com/apache/incubator-druid/pull/7038)).

But I hope that adding this nudge will help people including myself to adapt this practice consistently, because I think it's very powerful.

Reply:
Or rather, I would do what @clintropolis suggests here: https://github.com/apache/incubator-druid/pull/7206#discussion_r264400705 because it's elegant and promotes other good practices, apart from self-review.

Reply:
I would prefer if this review checklist was removed from this PR and then potentially re-added separately from the act of creating a pull request template. This is because if we do end up having an official guide / checklist it should be in the main Druid repo, rather than a separate repo. It will make it easier to discuss its specific contents and evolve it over time in a way that reflects community consensus.

Reply:
+1 for having a checklist in our repository, so it makes sense to me to add the checklist in a follow-up PR.

Reply:
Having two *almost identical* checklists in two different places will confuse people. Somebody may accidentally use the wrong checklist. People won't be able to see the difference easily.

I think it's much better to keep in the repository an auxiliary list of Druid-specific *overrides and amendments* to the generic checklist. It will additionally help to draw more attention to the smaller list of Druid-specific rules, which are doubly important if they *override* the generic rules.

I can create a document that for now just references the generic checklist and says

```
Use that generic checklist (link). This document will list
overrides and amendments to the generic checklist.

Currently, it's empty, nobody has added a Druid specific rule yet.
```

And link this document from the PR template.

Reply:
If the checklist is to be part of the project's policy, then I don't think it's appropriate for it to reside outside of the project.

> Somebody may accidentally use the wrong checklist. People won't be able to see the difference easily.

My stance on that would be that we should make it clear that the in-project checklist is the authoritative one.

Overall, I agree with the comments on removing the checklist from this for now and adding it in as a separate PR for an in-project checklist.

Reply:
Another idea would be to just bring the full checklist into the Druid repo with this PR.

Reply:
> If the checklist is to be part of the project's policy, then I don't think it's appropriate for it to reside outside of the project.

Why?

> My stance on that would be that we should make it clear that the in-project checklist is the authoritative one.

I can add the corresponding language to the in-repo document, for example:
```
Druid's checklist for concurrency consists of the items from
the general checklist (link), plus the following overrides and
amendments. In case of discrepancy, the items in this document
are more authoritative.

...
```

> Another idea would be to just bring the full checklist into the Druid repo with this PR.

There are multiple reasons why I don't want to do that. I currently want to reformat the checklist and create a static website a-la http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines, to be more digestable during real code reviews. Doing this within Druid repo is hardly possible.

Reply:
> Why?

If the checklist is meant to be part of the official policy applied to PR submissions, then having the checklist hosted on a non-Druid, non-Apache site (Medium), under the control of a single individual, effectively sets up an external individual authority over part of the project's policy. 

I trust that you wouldn't abuse that situation, but that's just not a good structure for official rules. Imagine if a national law had a section that said, "Refer to `<private individual's mutable blog post>` for details."

How would I revise the checklist if needed? The dependence on Medium seems unnecessary.

If the response to that is we can add "diffs" to the base checklist, I think that's problematic for the reasons @gianm mentions in https://github.com/apache/incubator-druid/pull/7206#discussion_r266679338, and still doesn't entirely do away with the issue of "external authority".

If the checklist is meant to be just recommended material (i.e., the instructions don't require the user to go through the checklist), then the Medium blog post is fine.

> There are multiple reasons why I don't want to do that. I currently want to reformat the checklist and create a static website a-la http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines, to be more digestable during real code reviews. Doing this within Druid repo is hardly possible.

Then I think the right approach would be to host it on the Druid site somewhere.

Reply:
Ok, I see the point now.

Now we are not talking about a blog post, we are talking about a repo. I can add all Druid committers to the list of committers of that repo as well. Would that work?

> Then I think the right approach would be to host it on the Druid site somewhere.

Unfortunately, practically this won't work because I would likely need to have something like a custom Jekyll config. So the checklist needs to be a separate repo. I'm open to moving the repo to `apache` org, but it doesn't seem to me that it may easily happen.

Reply:
I think that would be better, but I'm not really seeing why the content of the checklist couldn't be formatted into a standard Druid doc page.

Frankly, I wouldn't care about this aspect if we weren't under Apache, but now I feel we should be as "safe" as possible re: official policies and avoid anything that could be seen as even remotely questionable.

Maybe it would be good to bring this specific question up on the dev list and see what the mentors think.

Reply:
I don't see this as even remotely questionable. The project depends (with various meanings attached to "depends") on many non-Apache things, outside of any control by Druid committers, such as TeamCity, Travis, Github (all proprietary products), and dozens of non-Apache library dependencies. All these things seem to be "stronger" than the checklist that is not enforced in any way because ultimately committers review code themselves.

For example, a codebase already has several links to blog posts at https://shipilev.net as a "source of truth" explaining some things, you can say as well that Aleksey Shipilev can change the contents of those pages and they start to make some malicious advice, therefore we need to copy over the contents of that blog posts to Druid repo. This sounds absurd to me.

There are projects that are notourious for their extreme "self-containedness" and obsession to make an impression that they are the only thing in the world that exists, but I don't want Druid to be like that.

Reply:
> The project depends (with various meanings attached to "depends") on many non-Apache things, outside of any control by Druid committers, such as TeamCity, Travis, Github (all proprietary products), and dozens of non-Apache library dependencies. 

What would be the alternative to using TeamCity, Travis, Github, or various dependencies? It would be an immense amount of effort to reimplement those tools or libraries. 

Here, we would simply need to reformat some text and make a new doc page. I don't see sufficient justification for the indirection and complexity of having part of the project's official policies reside in an external repo.

> All these dependencies seem to be "stronger" than the checklist that is not enforced in any way because ultimately committers review code themselves.

Can we rephrase the template to make the checklist recommended reading, and not make it sound like an official requirement then? If we do that, then I think it's fine to not bring the checklist into the project.

> For example, a codebase already has several links to blog posts at https://shipilev.net as a "source of truth" explaining some things

I searched the code for that and see references in BitmapOffset and EventReceiverFirehoseFactory, and those simply provide context and technical explanation, they're not rules governing the committers or contributors.

Reply:
> What would be the alternative to using TeamCity, Travis, Github, or various dependencies? It would be an immense amount of effort to reimplement those tools or libraries.

We are not doing not because this would be an immense amount of effort, but because this is simply not needed.

> Here, we would simply need to reformat some text and make a new doc page. I don't see sufficient justification for the indirection and complexity of having part of the project's official policies reside in an external repo.

I already explained why practically the checklist should be in a separate repo on its own. In the current form, it can be wherever, but the current form is not digestible for reviews. Making it more easily navigable likely requires some Jekyll magic that is not compatible with Druid repo.

Reply:
What are your thoughts on positioning the checklist as a recommended resource instead of as part of the official requirements for the PR submission process? 

I'm fine with having it in an external repo in that case.

Reply:
> The project depends (with various meanings attached to "depends") on many non-Apache things, outside of any control by Druid committers, such as TeamCity, Travis, Github (all proprietary products), and dozens of non-Apache library dependencies. All these dependencies seem to be "stronger" than the checklist that is not enforced in any way because ultimately committers review code themselves.

While the project doesn't entirely control external tools/infra, the project does control in-repo what tools like TeamCity and Travis do. The inspection profiles, style and formatting definitions, test suites, test setup scripts, etc. are in-repo.

I'm not arguing against having external dependencies in general, just specifically in the area of formal rules that govern the community, out of a belief that we should try to keep the project's governance as "clean" and self-contained as possible.

As written, the PR makes the checklist sound very official, and I don't think that concerns about confusion arising from having a fork of the checklist or formatting difficulties are sufficient justification for having an external dependency in this area.

Reply:
I'll reword to make it recommended, not obligatory checklist.